service/migration_manager.cc

```diff
@@ -902,7 +902,7 @@ future<> migration_manager::push_schema_mutation(const gms::inet_address& endpoi
     return _messaging.send_definitions_update(id, std::move(fm), std::move(cm));
 }
 
-future<> migration_manager::announce_with_raft(std::vector<mutation> schema, group0_guard guard, std::string_view description) {
+future<> migration_manager::announce_with_raft(std::vector<mutation> schema, group0_guard guard, const sstring& description) {
```

Comment:
I think you've opened a huge pandora's box here.
First, there is a significant risk that nobody will remember in the future why you you changed to const sstring& here, and will change it back to std::string_view and nobody will notice this.
Second, this problem isn't at all specific to this specific functions - in many many functions all across the code, we've been slowly changing "const sstring&" to "std::string_view". All of those may be accidentally used with a temporary and risk this bug. I'm worried that maybe we have a dozen other instances of this bug and don't know...
I see this is a known pitfall - https://www.foonathan.net/2017/03/string_view-temporary/ - and an attempt to solve this http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0936r0.pdf.

Reply:
I would say changing to `string_view` is what's opening the pandora's box.

I can leave a comment explaining why I use `const sstring&`.

Reply:
Hmm but wait a sec...

the article you sent made me realize - calling `f(format(...))` where `f` takes a `string_view` should actually be safe, because the lifetime of `format(...)` ends not at the moment of entering `f`, but at the end of the entire expression `f(format(...))`!

However does it also apply for expressions such as `co_await f(format(...))`? Will the `format(...)` be destroyed only after the `co_await` is evaluated, or can it be destroyed as soon as `f` returns?

Reply:
> However does it also apply for expressions such as co_await f(format(...))? Will the format(...) be destroyed only after the co_await is evaluated, or can it be destroyed as soon as f returns?

Well, `f(format(...))` is not a full expression inside `co_await f(format(...))` according to the definition of "full expression", so yes, the temporary should be kept alive.
See also:
https://stackoverflow.com/questions/45744673/c-full-expression-standard-wording
https://stackoverflow.com/questions/62620147/c-coroutines-temporaries-in-co-await-expressions

I'll remove this first commit then.

Reply:
> Hmm but wait a sec...
> 
> the article you sent made me realize - calling `f(format(...))` where `f` takes a `string_view` should actually be safe, because the lifetime of `format(...)` ends not at the moment of entering `f`, but at the end of the entire expression `f(format(...))`!

Right, I was also reading this article now and realized the same thing...

> However does it also apply for expressions such as `co_await f(format(...))`? Will the `format(...)` be destroyed only after the `co_await` is evaluated, or can it be destroyed as soon as `f` returns?

Right! The problem is not the co_await but the asynchronous function itself - migration_manager::announce_with_raft -  which returns a future, and takes a string_view.
It can't "save" this string_view after "returning" a future (i.e., after the first preemption point).

So maybe that should be our general rule (and documented somewhere?) - string_view are great in normal, synchronous functions, but should be avoided asynchronous functions.
Or not avoided - but used with care. same as references where we have "conventions" about this (passing a reference to an asynchronous function usually means you promise to hold this reference alive until the future is complete).

Reply:
>  It can't "save" this string_view after "returning" a future (i.e., after the first preemption point).

Apparently it's "saved" until the full `co_await` expression ends, so it should be kept alive (according to the standard - compilers sometimes do their own thing...) even across yields.

Reply:
> passing a reference to an asynchronous function usually means you promise to hold this reference alive until the future is complete

Same rule applies to views. Accepting `const sstring&` instead of a view won't protect you from a user which passes a temporary sstring object.